### PR TITLE
Python: fix build on Fedora/SUSE

### LINKED ISSUE #37171: Installation issue: python
### Steps to reproduce the issue

```console
$ spack spec -I python@3.10.10+zlib
Input spec
--------------------------------
 -   python@3.10.10+zlib

Concretized
--------------------------------
[+]  python@3.10.10%gcc@7.5.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline
+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,7d40923,f2fd060 arch=linux-opensuse15-zen
[+]      ^bzip2@1.0.8%gcc@7.5.0~debug~pic+shared build_system=generic arch=linux-opensuse15-zen
[+]          ^diffutils@3.9%gcc@7.5.0 build_system=autotools arch=linux-opensuse15-zen
[+]      ^expat@2.5.0%gcc@7.5.0+libbsd build_system=autotools arch=linux-opensuse15-zen
[+]          ^libbsd@0.11.7%gcc@7.5.0 build_system=autotools arch=linux-opensuse15-zen
[+]              ^libmd@1.0.4%gcc@7.5.0 build_system=autotools arch=linux-opensuse15-zen
[+]      ^gdbm@1.23%gcc@7.5.0 build_system=autotools arch=linux-opensuse15-zen
[+]      ^gettext@0.21.1%gcc@7.5.0+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=linux-open
suse15-zen
[+]          ^libiconv@1.17%gcc@7.5.0 build_system=autotools libs=shared,static arch=linux-opensuse15-zen
[+]          ^libxml2@2.10.3%gcc@7.5.0~python build_system=autotools arch=linux-opensuse15-zen
[+]          ^tar@1.34%gcc@7.5.0 build_system=autotools zip=pigz arch=linux-opensuse15-zen
[+]              ^pigz@2.7%gcc@7.5.0 build_system=makefile arch=linux-opensuse15-zen
[+]              ^zstd@1.5.5%gcc@7.5.0+programs build_system=makefile compression=none libs=shared,static arch=linux-o
pensuse15-zen
[+]      ^libffi@3.4.4%gcc@7.5.0 build_system=autotools arch=linux-opensuse15-zen
[+]      ^libxcrypt@4.4.33%gcc@7.5.0~obsolete_api build_system=autotools arch=linux-opensuse15-zen
[+]          ^perl@5.36.0%gcc@7.5.0+cpanm+open+shared+threads build_system=generic arch=linux-opensuse15-zen
[+]              ^berkeley-db@18.1.40%gcc@7.5.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux
-opensuse15-zen
[+]      ^ncurses@6.4%gcc@7.5.0~symlinks+termlib abi=none build_system=autotools arch=linux-opensuse15-zen
[+]      ^openssl@1.1.1t%gcc@7.5.0~docs~shared build_system=generic certs=mozilla arch=linux-opensuse15-zen
[+]          ^ca-certificates-mozilla@2023-01-10%gcc@7.5.0 build_system=generic arch=linux-opensuse15-zen
[+]      ^pkgconf@1.8.0%gcc@7.5.0 build_system=autotools arch=linux-opensuse15-zen
[+]      ^readline@8.2%gcc@7.5.0 build_system=autotools patches=bbf97f1 arch=linux-opensuse15-zen
[+]      ^sqlite@3.40.1%gcc@7.5.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=l
inux-opensuse15-zen
[+]      ^util-linux-uuid@2.38.1%gcc@7.5.0 build_system=autotools arch=linux-opensuse15-zen
[+]      ^xz@5.4.1%gcc@7.5.0~pic build_system=autotools libs=shared,static arch=linux-opensuse15-zen
[+]      ^zlib@1.2.13%gcc@7.5.0+optimize+pic+shared build_system=makefile arch=linux-opensuse15-zen

$ spack install --overwrite -y python@3.10.10

```

### Error message

Instance of python installed by spack cannot import the zlib module

<details><summary>Error message</summary>

<pre>
```console
❯ spack load python@3.10.10

❯ which python3
/home/awadell/spack/opt/spack/linux-opensuse15-zen/gcc-7.5.0/python-3.10.10-xplxbc5o5jrloglj35wngefohj7tt6ze/bin/python3

❯ python3 -c "import zlib"
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'zlib'

❯ PYTHONHOME=~/spack/opt/spack/linux-opensuse15-zen/gcc-7.5.0/python-3.10.10-xplxbc5o5jrloglj35wngefohj7tt6ze/ python3 -c "import zlib"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'zlib'
```
</pre></details>

### Information on your system

❯ spack debug report
* **Spack:** 0.20.0.dev0 (c76023d7accc53f40e49481179bdb26c8c18845a)
* **Python:** 3.6.13
* **Platform:** linux-opensuse15-zen3
* **Concretizer:** clingo

### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/11315222/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/11315224/spack-build-env.txt)
[spack-configure-args.txt](https://github.com/spack/spack/files/11315241/spack-configure-args.txt)
[spack-build-01-configure-out.txt](https://github.com/spack/spack/files/11315242/spack-build-01-configure-out.txt)

@adamjstewart @pradyunsg @scheibelp @skosukhin 

### Staged Version works

Installing with spack and keeping the stage, results in a usable instance, but the installed version lack's zlib:

```console
❯ spack install --overwrite --keep-stage python@3.10.10+zlib
...

❯ spack cd python@3.10.10+zlib

❯ ./python
Python 3.10.10 (main, Apr 24 2023, 19:39:34) [GCC 7.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import zlib
>>> 
```

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

### LINKED ISSUE #39551: Broken Python3 on OpenSUSE, lib-dynload not correctly linked
### Steps to reproduce the issue

Install py-pip (dependency of amdlibm) with AOCC@4.1.0:
```console
$ spack install py-pip %aocc
==> Warning: config:terminal_title has been replaced by install_status and is ignored
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/bzip2-1.0.8-q4pqsh3rxddsvyeuoc4umnhp4fbq3fcc
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/libmd-1.0.4-lp7x5tmamdnni6ccsick632krdghn4rl
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/ncurses-6.4-twp7yf3grnnzv3whbfr3fmccaml2s5ut
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/libiconv-1.17-xo2d4zaqgk4zq54eeic2sd6jxjhbxvuy
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/xz-5.4.1-huprvmsxzdvqjt4wiymsravg6ktpretm
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/zlib-ng-2.1.3-52s7eq32xbgfb2ecxnytevwzjfextkwo
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/zstd-1.5.5-knlbullsxnlliav3ee24osy35foxd26b
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/libffi-3.4.4-lqb7gtbxs7ugk2qj5z3eompbhiipra7k
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/libxcrypt-4.4.35-vyjg3panmbji2dy6oxo2iwvmvbm6tmr7
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/util-linux-uuid-2.38.1-p7rabtpzawmfnfxq5izkas34umhvtppj
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/libbsd-0.11.7-7cwjeplhcnehfrravoluexflx6rok44s
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/readline-8.2-so6tfuvqypsh7axma25ftc3kkroykzb4
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/pigz-2.7-rh55xrdkycsoipc7dlak6bgx2j3pmk3b
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/openssl-3.1.2-jufrpmzekubxplnjtioe3n4xri4zpjcr
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/libxml2-2.10.3-kcosc37y57aqexfzhvtjghvo2lxiwlve
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/expat-2.5.0-n3z6tlhhz3va2d2ympka6pz4jexp2sra
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/sqlite-3.42.0-ioreb5arslq7q7iy6m4bmyc6o4juiju4
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/gdbm-1.23-4rs5sb5jp4ohcu42kxqaqjcuijwm6wqe
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/tar-1.34-dztpp3xiyjxd4dthstdihajji24s7v4x
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/gettext-0.21.1-lvmu6twonklvtbhalkpccroswh6xtze2
[+] /home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/python-3.10.12-fvclniug6pdjl475gjvkkcfjmh4cpnyp
==> Installing py-pip-23.1.2-6j6wfgdu4xv4x6z5r56c4bdck3ekjs77 [22/22]
==> No binary for py-pip-23.1.2-6j6wfgdu4xv4x6z5r56c4bdck3ekjs77 found: installing from source
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
==> Using cached archive: /var/tmp/yanghaoyu/spack-cache/_source-cache/archive/3e/3ef6ac33239e4027d9a5598a381b9d30880a1477e50039db2eac6e8a8f6d1b18
==> No patches needed for py-pip
==> py-pip: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 1:
    '/home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/python-3.10.12-fvclniug6pdjl475gjvkkcfjmh4cpnyp/bin/python3.10' '/tmp/yanghaoyu/spack-stage/spack-stage-py-pip-23.1.2-6j6wfgdu4xv4x6z5r56c4bdck3ekjs77/spack-src/pip-23.1.2-py3-none-any.whl/pip' '-vvv' '--no-input' '--no-cache-dir' '--disable-pip-version-check' 'install' '--no-deps' '--ignore-installed' '--no-build-isolation' '--no-warn-script-location' '--no-index' '--prefix=/home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/py-pip-23.1.2-6j6wfgdu4xv4x6z5r56c4bdck3ekjs77' '/tmp/yanghaoyu/spack-stage/spack-stage-py-pip-23.1.2-6j6wfgdu4xv4x6z5r56c4bdck3ekjs77/spack-src/pip-23.1.2-py3-none-any.whl'
See build log for details:
  /tmp/yanghaoyu/spack-stage/spack-stage-py-pip-23.1.2-6j6wfgdu4xv4x6z5r56c4bdck3ekjs77/spack-build-out.txt
````

```console
$ spack spec -I py-pip
==> Warning: config:terminal_title has been replaced by install_status and is ignored
Input spec
--------------------------------
 -   py-pip

Concretized
--------------------------------
 -   py-pip@23.1.2%aocc@4.1.0 build_system=generic arch=linux-opensusetumbleweed2.38-zen3
[+]      ^python@3.10.12%aocc@4.1.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,7d40923,f2fd060 arch=linux-opensusetumbleweed2.38-zen3
[+]          ^bzip2@1.0.8%aocc@4.1.0~debug~pic+shared build_system=generic arch=linux-opensusetumbleweed2.38-zen3
[+]              ^diffutils@3.9%aocc@4.1.0 build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]          ^expat@2.5.0%aocc@4.1.0+libbsd build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]              ^libbsd@0.11.7%aocc@4.1.0 build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]                  ^libmd@1.0.4%aocc@4.1.0 build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]          ^gdbm@1.23%aocc@4.1.0 build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]          ^gettext@0.21.1%aocc@4.1.0+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]              ^libiconv@1.17%aocc@4.1.0 build_system=autotools libs=shared,static arch=linux-opensusetumbleweed2.38-zen3
[+]              ^libxml2@2.10.3%aocc@4.1.0~python build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]              ^tar@1.34%aocc@4.1.0 build_system=autotools zip=pigz arch=linux-opensusetumbleweed2.38-zen3
[+]                  ^pigz@2.7%aocc@4.1.0 build_system=makefile arch=linux-opensusetumbleweed2.38-zen3
[+]                  ^zstd@1.5.5%aocc@4.1.0+programs build_system=makefile compression=none libs=shared,static arch=linux-opensusetumbleweed2.38-zen3
[+]          ^libffi@3.4.4%aocc@4.1.0 build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]          ^libxcrypt@4.4.35%aocc@4.1.0~obsolete_api build_system=autotools patches=4885da3 arch=linux-opensusetumbleweed2.38-zen3
[+]              ^perl@5.38.0%aocc@4.1.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-opensusetumbleweed2.38-zen3
[+]                  ^berkeley-db@18.1.40%aocc@4.1.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-opensusetumbleweed2.38-zen3
[+]          ^ncurses@6.4%aocc@4.1.0~symlinks+termlib abi=none build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]          ^openssl@3.1.2%aocc@4.1.0~docs+shared build_system=generic certs=mozilla arch=linux-opensusetumbleweed2.38-zen3
[+]              ^ca-certificates-mozilla@2023-05-30%aocc@4.1.0 build_system=generic arch=linux-opensusetumbleweed2.38-zen3
[+]          ^pkgconf@1.9.5%aocc@4.1.0 build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]          ^readline@8.2%aocc@4.1.0 build_system=autotools patches=bbf97f1 arch=linux-opensusetumbleweed2.38-zen3
[+]          ^sqlite@3.42.0%aocc@4.1.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]          ^util-linux-uuid@2.38.1%aocc@4.1.0 build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]          ^xz@5.4.1%aocc@4.1.0~pic build_system=autotools libs=shared,static arch=linux-opensusetumbleweed2.38-zen3
[+]          ^zlib-ng@2.1.3%aocc@4.1.0+compat+opt build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
```

```console
$ spack spec -I python
==> Warning: config:terminal_title has been replaced by install_status and is ignored
Input spec
--------------------------------
 -   python

Concretized
--------------------------------
[+]  python@3.10.12%aocc@4.1.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,7d40923,f2fd060 arch=linux-opensusetumbleweed2.38-zen3
[+]      ^bzip2@1.0.8%aocc@4.1.0~debug~pic+shared build_system=generic arch=linux-opensusetumbleweed2.38-zen3
[+]          ^diffutils@3.9%aocc@4.1.0 build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]      ^expat@2.5.0%aocc@4.1.0+libbsd build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]          ^libbsd@0.11.7%aocc@4.1.0 build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]              ^libmd@1.0.4%aocc@4.1.0 build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]      ^gdbm@1.23%aocc@4.1.0 build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]      ^gettext@0.21.1%aocc@4.1.0+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]          ^libiconv@1.17%aocc@4.1.0 build_system=autotools libs=shared,static arch=linux-opensusetumbleweed2.38-zen3
[+]          ^libxml2@2.10.3%aocc@4.1.0~python build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]          ^tar@1.34%aocc@4.1.0 build_system=autotools zip=pigz arch=linux-opensusetumbleweed2.38-zen3
[+]              ^pigz@2.7%aocc@4.1.0 build_system=makefile arch=linux-opensusetumbleweed2.38-zen3
[+]              ^zstd@1.5.5%aocc@4.1.0+programs build_system=makefile compression=none libs=shared,static arch=linux-opensusetumbleweed2.38-zen3
[+]      ^libffi@3.4.4%aocc@4.1.0 build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]      ^libxcrypt@4.4.35%aocc@4.1.0~obsolete_api build_system=autotools patches=4885da3 arch=linux-opensusetumbleweed2.38-zen3
[+]          ^perl@5.38.0%aocc@4.1.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-opensusetumbleweed2.38-zen3
[+]              ^berkeley-db@18.1.40%aocc@4.1.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-opensusetumbleweed2.38-zen3
[+]      ^ncurses@6.4%aocc@4.1.0~symlinks+termlib abi=none build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]      ^openssl@3.1.2%aocc@4.1.0~docs+shared build_system=generic certs=mozilla arch=linux-opensusetumbleweed2.38-zen3
[+]          ^ca-certificates-mozilla@2023-05-30%aocc@4.1.0 build_system=generic arch=linux-opensusetumbleweed2.38-zen3
[+]      ^pkgconf@1.9.5%aocc@4.1.0 build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]      ^readline@8.2%aocc@4.1.0 build_system=autotools patches=bbf97f1 arch=linux-opensusetumbleweed2.38-zen3
[+]      ^sqlite@3.42.0%aocc@4.1.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]      ^util-linux-uuid@2.38.1%aocc@4.1.0 build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
[+]      ^xz@5.4.1%aocc@4.1.0~pic build_system=autotools libs=shared,static arch=linux-opensusetumbleweed2.38-zen3
[+]      ^zlib-ng@2.1.3%aocc@4.1.0+compat+opt build_system=autotools arch=linux-opensusetumbleweed2.38-zen3
```

### Error message

<details>
<summary>Error message</summary>
<pre>
==> Installing py-pip-23.1.2-6j6wfgdu4xv4x6z5r56c4bdck3ekjs77 [22/22]
==> No binary for py-pip-23.1.2-6j6wfgdu4xv4x6z5r56c4bdck3ekjs77 found: installing from source
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
==> Using cached archive: /var/tmp/yanghaoyu/spack-cache/_source-cache/archive/3e/3ef6ac33239e4027d9a5598a381b9d30880a1477e50039db2eac6e8a8f6d1b18
==> No patches needed for py-pip
==> py-pip: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 1:
    '/home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/python-3.10.12-fvclniug6pdjl475gjvkkcfjmh4cpnyp/bin/python3.10' '/tmp/yanghaoyu/spack-stage/spack-stage-py-pip-23.1.2-6j6wfgdu4xv4x6z5r56c4bdck3ekjs77/spack-src/pip-23.1.2-py3-none-any.whl/pip' '-vvv' '--no-input' '--no-cache-dir' '--disable-pip-version-check' 'install' '--no-deps' '--ignore-installed' '--no-build-isolation' '--no-warn-script-location' '--no-index' '--prefix=/home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/py-pip-23.1.2-6j6wfgdu4xv4x6z5r56c4bdck3ekjs77' '/tmp/yanghaoyu/spack-stage/spack-stage-py-pip-23.1.2-6j6wfgdu4xv4x6z5r56c4bdck3ekjs77/spack-src/pip-23.1.2-py3-none-any.whl'
See build log for details:
  /tmp/yanghaoyu/spack-stage/spack-stage-py-pip-23.1.2-6j6wfgdu4xv4x6z5r56c4bdck3ekjs77/spack-build-out.txt
</pre></details>

<details>
<summary>Build Log</summary>

<pre>
=> py-pip: Executing phase: 'install'
==> [2023-08-21-11:19:51.647349] '/home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/python-3.10.12-fvclniug6pdjl475gjvkkcfjmh4cpnyp/bin/python3.10' '/tmp/yanghaoyu/spack-stage/spack-stage-py-pip-23.1.2-6j6wfgdu4xv4x6z5r56c4bdck3ekjs77/spack-src/pip-23.1.2-py3-none-any.whl/pip' '-vvv' '--no-input' '--no-cache-dir' '--disable-pip-version-check' 'install' '--no-deps' '--ignore-installed' '--no-build-isolation' '--no-warn-script-location' '--no-index' '--prefix=/home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/py-pip-23.1.2-6j6wfgdu4xv4x6z5r56c4bdck3ekjs77' '/tmp/yanghaoyu/spack-stage/spack-stage-py-pip-23.1.2-6j6wfgdu4xv4x6z5r56c4bdck3ekjs77/spack-src/pip-23.1.2-py3-none-any.whl'
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Traceback (most recent call last):
  File "<frozen zipimport>", line 577, in _get_decompress_func
ModuleNotFoundError: No module named 'zlib'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<frozen zipimport>", line 625, in _get_data
  File "<frozen zipimport>", line 580, in _get_decompress_func
zipimport.ZipImportError: can't decompress data; zlib not available

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/python-3.10.12-fvclniug6pdjl475gjvkkcfjmh4cpnyp/lib/python3.10/runpy.py", line 157, in _get_module_details
    code = loader.get_code(mod_name)
  File "<frozen zipimport>", line 196, in get_code
  File "<frozen zipimport>", line 759, in _get_module_code
  File "<frozen zipimport>", line 627, in _get_data
zipimport.ZipImportError: can't decompress data; zlib not available

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/python-3.10.12-fvclniug6pdjl475gjvkkcfjmh4cpnyp/lib/python3.10/runpy.py", line 189, in _run_module_as_main
    mod_name, mod_spec, code = _get_main_module_details(_Error)
  File "/home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/python-3.10.12-fvclniug6pdjl475gjvkkcfjmh4cpnyp/lib/python3.10/runpy.py", line 238, in _get_main_module_details
    return _get_module_details(main_name)
  File "/home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/python-3.10.12-fvclniug6pdjl475gjvkkcfjmh4cpnyp/lib/python3.10/runpy.py", line 159, in _get_module_details
    raise error(format(e)) from e
ImportError: can't decompress data; zlib not available
</pre></details>

<details>
<summary>Running Python3 Directly</summary>
<pre>
$ which python3
/home/yanghaoyu/opt/spack/packages/linux-opensusetumbleweed2.38-zen3/aocc-4.1.0/python-3.10.12-fvclniug6pdjl475gjvkkcfjmh4cpnyp/bin/python3. 
$ python3
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Python 3.10.12 (main, Aug 21 2023, 10:11:54) [Clang 16.0.3 (CLANG: AOCC_4.1.0-Build#270 2023_07_10)] on linux
Type "help", "copyright", "credits" or "license" for more information.
Traceback (most recent call last):
  File "/etc/pythonstart", line 7, in <module>
    import readline
ModuleNotFoundError: No module named 'readline'
</pre></details>

### Information on your system

```
$ spack debug report:
* **Spack:** 0.21.0.dev0 (c9ec5fb9ac565033c7db643f0b51da5dc9315ac0)
* **Python:** 3.11.4
* **Platform:** linux-opensusetumbleweed2.38-zen3
==> Warning: config:terminal_title has been replaced by install_status and is ignored
* **Concretizer:** clingo
```

### Additional information

Python maintainers:
@adamjstewart  @scheibelp  @skosukhin

py-pip maintainers:
@adamjstewart  @pradyunsg

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by adamjstewart**
Fixes #37171 
Fixes #39551

@DanielYang59

### CODE REVIEW

File: var/spack/repos/builtin/packages/python/package.py [hunk 1]
```diff
@@ -542,6 +542,10 @@ def configure_args(self):
         config_args = []
         cflags = []
 
+        # https://bugs.python.org/issue1294959
+        if "fedora" in self.spec.os or "suse" in self.spec.os:
```
Comment (haampie) [inline var/spack/repos/builtin/packages/python/package.py:546]: is it possible to do this without `spec.os`? I'd like to be able to drop that property. For spack it makes no sense at all to have a lib64 directory :confused: can't you just drop the branch and force `lib` everywhere?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/python/package.py:546]: Lib is already the default but doesn't seem to work on Fedora/SUSE, I think it needs to be lib64. We can do it without `spec.os`, I just need some way of checking to see which distro we're on.

Comment (haampie) [reply var/spack/repos/builtin/packages/python/package.py:546]: What makes it such that it needs to be lib64? Something in Python itself? Where?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/python/package.py:546]: No idea, try reading through https://bugs.python.org/issue1294959 and see if you can extract more info than me.

**Comment by michaelkuhn** [comment]
I cannot reproduce the original problem on Fedora:
```
$ spack load python+zlib   
$ python -c 'import zlib'
$ ls $(spack location -i python+zlib)
bin/  include/  lib/  share/
```
Am I missing anything?

**Comment by DanielYang59** [comment]
> I cannot reproduce the original problem on Fedora:
> 
> ```
> $ spack load python+zlib   
> $ python -c 'import zlib'
> $ ls $(spack location -i python+zlib)
> bin/  include/  lib/  share/
> ```
> 
> Am I missing anything?

@michaelkuhn May be specific to OpenSUSE Tumbleweed?

**Review by manuelakuhn — COMMENTED** [review]
I tried to reproduce the problems on my Fedora and inside an opensuse/tumbleweed container but could not reproduce it in either of them.
I tried 
```
# spack load python+zlib
# python -c 'import zlib'
# ls $(spack location -i python+zlib)
.spack  bin  include  lib  share
```
and `spack install py-pip ^zlib` which ran through without problems.

**Comment by adamjstewart** [comment]
@DanielYang59 @awadell1 any other ideas for reproducing the original issue?

**Comment by DanielYang59** [comment]
@adamjstewart I'm working on reproducing this issue on my PC (with Zen 4 CPU) on a OpenSUSE Tumbleweed VM. However I cannot even install the berkeley-db dependency and got `checking whether the C+= compiler supports templates for STL... configure: error: no` error (gcc-c++-13 is installed on my VM). I'm working on resolving this and would keep you updated.

**Comment by michaelkuhn** [comment]
I also gave this another try in a Tumbleweed container and wasn't able to reproduce the issue. `python+zlib`'s prefix contains a `lib` directory, the zlib module imports fine and `py-pip` can also be installed.

**Comment by adamjstewart** [comment]
Closing until we find a way to reproduce the original issues.

**Comment by DanielYang59** [comment]
@HadrienG2 is able to reproduce this on a local install (not a container) just as my original case in https://github.com/spack/spack/issues/39551#issuecomment-1894193652.
> I managed to reproduce this, but only on a local openSUSE Tumbleweed install, not on an opensuse/tumbleweed container. No idea what is different between the containerized and non-containerized version though... 🤷

Can you please follow up this please? Sorry I cannot help much on this as I wiped my local OpenSUSE image.

**Comment by vicentebolea** [comment]
As a both OpenSUSE and spack user I still see this error whenever I install python in spack. I would like to help with this

**Comment by adamjstewart** [comment]
Can you open a new PR? Or comment on the open issue? I think someone other than me should spearhead this fix since I don't have access to these systems. Feel free to steal my changes for the new PR, I don't care about credit.

File: var/spack/repos/builtin/packages/python/package.py (hunk 1 above)
Comment (vicentebolea) [reply var/spack/repos/builtin/packages/python/package.py:546]: In the case of Opensuse the issue here is that the autoconf installed by Opensuse defaults to install libs in lib64, is there a way to conflict with external package. In this manner we can use the spack autoconf.

**Comment by vicentebolea** [comment]
Followup PR at: https://github.com/spack/spack/pull/48575

**Pull request closed without merging**